PolygonBrush

rvankevelaer opened the pull request:
Fehler:

1. ~~First and last coordinate are not equivalent in ModifySubtract~~
2. ~~"Verspringen/Einbruch" von Koordinaten bei ModifySubtract (ggf. durch Auskommentieren behoben (?))~~
3. ~~Komplettes Löschen scheint nicht möglich zu sein - Vielleicht nur ein Fehler bei der Aktualisierung?~~

Commit:
First version of PolygonBrush. Circle as sketchPoint, radius manipulation using mouse wheel, drawing circles in the size of the sketchPoint.

Commit:
Drawing intersections of two polygons works

Commit:
Intersecting polygons are merged, holes closed.

Commit:
Polygon drawing on mouseover and merging with existing annotations working

Commit:
Union and difference working

Commit:
Erase function can divide and remove existing annotations

Commit:
Error handling, dividing polygons by erase function

Commit:
Polygon Add and Subtract in two different interactions

Commit:
PolygonAdd/Brush/Subtract Code Cleanup

Commit:
PolygonSubtract needs less memory now

Commit:
PolygonAdd uses booleanOverlap instead of intersects

Commit:
Add/Subtract without intersection objects

Commit:
Prepared new modify interactions, standalone PolygonAdd just adding polygons without modification

Commit:
ModifyAdd interaction work in progress

Commit:
ModifyAdd first working version

Commit:
Fast solution for PolygonAdd, less new objects

Commit:
Circle visible when editing, most functions inherited from superclass

Commit:
First working version of ModifySubtract, helper function file for union and difference

Commit:
If a polygon is cut in two pieces by ModifySubtract, the larger piece is kept on the map

Commit:
No start/finish drawing in ModifySubtract at PointerUpEvent

Commit:
Project cleaned up

Commit:
Removed unused superclass PolygonBrush, new finishDrawing functions, renamed functions and variables, code cleanup.

mzur committed:
Rename PolygonDraw interaction to PolygonBrush

mzur committed:
Merge ModifyAdd and ModifySubtract interactions to ModifyPolygonBrush

Add and subtract is controlled with the new "mode" option.

mzur committed:
Split polygonInteractionHelpers into separate files

mzur committed:
Remove unneeded @turf/area and @turf/intersect

mzur commented on src/ol/interaction/ModifyPolygonBrush.js:
```diff
@@ -0,0 +1,283 @@
+
+    this.sketchPoint_ = null;
+
+    this.mode_ = options.mode ? options.mode : 'add';
```
Vielleicht könnte man hier statt des `mode` auch so etwas wie eine `addCondition` und `subtractCondition` einbauen. Das gibt es bei der normalen [Modify Interaction](https://openlayers.org/en/latest/apidoc/module-ol_interaction_Modify-Modify.html) auch und kann so etwas sein wie ["linke Maustaste", "rechte Maustaste" oder auch "nichts"](https://openlayers.org/en/latest/apidoc/module-ol_events_condition.html#~Condition). Damit wäre die Interaction sehr flexibel. Man könnte sowohl die Idee mit "linke Maustaste=add", "rechte Maustaste=subtract" umsetzen, als auch die Möglichkeit für beides separate Buttons im User Interface zu haben.

mzur commented on src/ol/geom/flat/difference.js:
```diff
@@ -0,0 +1,28 @@
+import Feature from '../../Feature.js';
+import Polygon from '../Polygon.js';
+
+//order of polygons is important
```
Kommentar entweder genauer ausführen oder einfach weglassen. So ist er nicht gut zu verstehen.

mzur commented on src/ol/geom/flat/difference.js:
```diff
@@ -0,0 +1,28 @@
+import Polygon from '../Polygon.js';
+
+//order of polygons is important
+export function difference(comparePolygon,currentPolygon) {
```
Leerzeichen hinter dem Komma. Das gilt überall.

mzur commented on src/ol/geom/flat/difference.js:
```diff
@@ -0,0 +1,28 @@
+            var currentPolygon = turfPolygon(differencePolygon.geometry.coordinates[i]);
+            var olPoly = new Feature(new Polygon(differencePolygon.geometry.coordinates[i]))
+            var area = olPoly.getGeometry().getArea()
+//            console.log("OL Poly area:",area,"Max area:",maxArea)
```
Bitte alle Kommentare löschen, die nicht gebraucht werden oder zum Debugging gehörten.

mzur commented on src/ol/geom/flat/difference.js:
```diff
@@ -0,0 +1,28 @@
+        var maxPoly;
+        for (var i = 0; i < differencePolygon.geometry.coordinates.length; i++) {
+            var currentPolygon = turfPolygon(differencePolygon.geometry.coordinates[i]);
+            var olPoly = new Feature(new Polygon(differencePolygon.geometry.coordinates[i]))
```
Bitte alle entsprechenden Zeilen mit `;` abschließen.

mzur commented on src/ol/geom/flat/difference.js:
```diff
@@ -0,0 +1,28 @@
+//        console.log("Result:",maxOL);
+    }
+    var coords = differencePolygon.geometry.coordinates;
+    return coords;
```
Viele Coding Style Guides sehen es vor, dass man vor einem `return` eine Leerzeile lässt. Darauf würde ich jetzt aber nicht bestehen.

mzur commented on src/ol/interaction/ModifyPolygonBrush.js:
```diff
@@ -0,0 +1,283 @@
+     */
+    this.lastPointerEvent_ = null;
+
+    this.circleRadius_ = 10000;  //TODO better value
```
Alle TODOs müssen noch aufgelöst werden. In diesem Fall hatten wir glaube ich vorgesehen, dass man den Radius mit Shift+Mausrad anpassen kann (oder?). Außerdem sollte der Radius auch relativ zur aktuellen Zoomstufe der Karte sein. D.h. wenn man rein- oder rauszoomt, dann sollte der Kreis immer die selbe (scheinbare) Größe behalten.

mzur commented on src/ol/interaction/ModifyPolygonBrush.js:
```diff
@@ -0,0 +1,283 @@
+    if (this.sketchPoint_) {
+      const sketchPointGeom = /** @type {Circle} */ (this.sketchPoint_.getGeometry());
+      if (event.originalEvent.deltaY > 0) {
+        this.circleRadius_ = sketchPointGeom.getRadius() + 1000;  //TODO better value
```
S.o. zum `circleRadius_`.

mzur commented on src/ol/interaction/ModifyPolygonBrush.js:
```diff
@@ -0,0 +1,283 @@
+        this.circleRadius_ = sketchPointGeom.getRadius() + 1000;  //TODO better value
+      }
+      if (event.originalEvent.deltaY < 0) {
+        this.circleRadius_ = sketchPointGeom.getRadius() - 1000;  //TODO better value
```
S.o. zum `circleRadius_`.

mzur commented on src/ol/interaction/ModifyPolygonBrush.js:
```diff
@@ -0,0 +1,283 @@
+      pass = false;
+      this.updateSketchPointRadius_(event);
+    }
+    if (btn == 0 && (type === MapBrowserEventType.POINTERDOWN)) {
```
In den meisten Fällen ist es besser, wenn man `===` statt `==` verwendet. Das würde ich hier auch so halten.

mzur commented on src/ol/interaction/ModifyPolygonBrush.js:
```diff
@@ -0,0 +1,283 @@
+
+    var currentPolygon = turfPolygon(geometry.getCoordinates())
+
+    //TODO dispatch event here or in the handler above? Then an event must be passed!
```
Das Event (`modifyend`, nicht `drawend`) soll nur einmal gefeuert werden, wenn die Maustaste losgelassen wird.

mzur commented on src/ol/interaction/ModifyPolygonBrush.js:
```diff
@@ -0,0 +1,283 @@
+    // First dispatch event to allow full set up of feature
+//    this.dispatchEvent(new DrawEvent(DrawEventType.DRAWEND, sketchFeature));
+
+    // Then insert feature
```
Ich weiß nicht, ob dieser Kommentar noch vom original Code stammt. Solche Kommentare sind aber eher nutzlos. Man sollte eher nicht beschreiben "was" gemacht wird (das macht ja schon der Code selbst), sondern eher "warum" etwas gemacht wird.

mzur commented on src/ol/interaction/ModifyPolygonBrush.js:
```diff
@@ -0,0 +1,283 @@
+      this.source_.addFeature(sketchFeature);
+    }
+
+    //TODO skip changes if modifyFeature_ contains sketchFeature_ other completely
```
Dieser Kommentar scheint nichts mit dem `if ... else` zu tun zu haben.

mzur commented on src/ol/interaction/PolygonBrush.js:
```diff
@@ -0,0 +1,146 @@
+  constructor(options) {
+
+    super(options)
+    this.circleRadius_ = 10000  //TODO better value
```
Dies sollte genauso behandelt werden wie bei ModifyPolygonBrush.

mzur commented on src/ol/interaction/PolygonBrush.js:
```diff
@@ -0,0 +1,146 @@
+      pass = false;
+      this.updateSketchPointRadius_(event);
+    }
+    if (btn == 0 && (type === MapBrowserEventType.POINTERDOWN)) {
```
Ist die Unterscheidung zwischen den Buttons hier noch notwendig?

mzur commented on src/ol/interaction/PolygonBrush.js:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  handleUpEvent(event) {
+    //this is overridden because nothing should happen here
```
Sollte hier nicht das `drawend` Event gefeuert werden? Oder wird das im Original Code auch alles in `handleEvent` gemacht?

mzur replied:
Habe das Event jetzt entsprechend hinzugefügt.

mzur commented on src/ol/interaction/PolygonBrush.js:
```diff
@@ -0,0 +1,146 @@
+    if (this.sketchPoint_) {
+      const sketchPointGeom = /** @type {Circle} */ (this.sketchPoint_.getGeometry());
+      if (event.originalEvent.deltaY > 0) {
+        this.circleRadius_ = sketchPointGeom.getRadius() + 1000;  //TODO better value
```
S.o.

mzur commented on src/ol/interaction/PolygonBrush.js:
```diff
@@ -0,0 +1,146 @@
+        this.circleRadius_ = sketchPointGeom.getRadius() + 1000;  //TODO better value
+      }
+      if (event.originalEvent.deltaY < 0) {
+        this.circleRadius_ = sketchPointGeom.getRadius() - 1000;  //TODO better value
```
S.o.

mzur reviewed (changes requested):
Ich habe die Interactions umbenannt und die beiden Modify Interactions zu einer einzigen zusammengefasst. Einige meiner Kommentare betreffen den Coding Style. Da musst du teilweise für dich selbst entscheiden, was du für deinen "Stil" für sinnvoll hältst. Die größte Sache, die noch implementiert werden sollte, ist wahrscheinlich der dynamische Radius des Kreises.

Falls du möchtest, kannst du auch einmal [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/) lesen. Die Regeln sind schnell verinnerlicht und verbessern die Lesbarkeit von Commits enorm, wie ich finde.

Die drei Bugs, über die wir gesprochen haben, schaue ich mir jetzt noch an.

mzur committed:
Simplify PolygonBrush to investigate artifact issue

mzur committed:
Fix stopping of event when sketch point radius is changed

mzur committed:
Fix coordinate precision issue with union operation

mzur commented:
Ich denke, ich habe den Fehler mit dem "Einbruch" bei der Draw Interaction gelöst (
4f50d7e). Hat mich auch nur ein paar Stunden gekostet :wink: Wahrscheinlich lässt sich das bei der Subtract Interaction genauso lösen.

mzur committed:
Fix coordinate precision issue with difference operation

mzur commented:
Scheinkt geklappt zu haben. Vielleicht ist Punkt 1 damit auch behoben. @rvankevelaer schau mal, ob du das noch reproduzieren kannst.

Commit:
Review code style

- Added missing ';' to line ends
- Added space after commas
- Removed unneeded comments
- Changed all '==' to '==='

Commit:
Implement resolution-dependent setting of sketchPointRadius

Commit:
Implement missing Add case in ModifyPolygonBrush

- If the sketchFeature contains the modifyFeature completely,
  the modifyFeature gets the sketchFeature's coordinates.
- Improved variable names for better code readability
- Remove unneeded comments

Commit:
Remove additional case EventType === MOUSEWHEEL in handleEvent

Commit:
Use separate conditions for add and subtract instead of one mode variable in ModifyPolygonBrush.js

ghost commented:
> Scheinkt geklappt zu haben. Vielleicht ist Punkt 1 damit auch behoben. @rvankevelaer schau mal, ob du das noch reproduzieren kannst.

Bisher ist das Problem nicht wieder vorgekommen.

ghost commented:
> Fehler:
> 
>     1. ~First and last coordinate are not equivalent in ModifySubtract~
> 
>     2. ~"Verspringen/Einbruch" von Koordinaten bei ModifySubtract (ggf. durch Auskommentieren behoben (?))~
> 
>     3. Komplettes Löschen scheint nicht möglich zu sein - Vielleicht nur ein Fehler bei der Aktualisierung?

Komplettes Löschen ist noch immer nicht möglich in der ModifyPolygonBrush. Es sieht allerdings so aus, dass sowohl das Feature-Array als auch das Feature-Array der VectorSource nach dem Löschen der Features leer sind.

Vielleicht handelt es sich auch um ein Problem bei der Aktualisierung der Map?

Commit:
Implement MODIFYREMOVE signal

mzur committed:
Simplify PolygonBrush

mzur committed:
Simplify ModifyPolygonBrush interaction

mzur committed:
Fix smooth updating of sketch feature radius in PolygonBrush

mzur committed:
Fix contained case for ModifyPolygonBrush

mzur reviewed (changes requested):
Notiz an mich: Das Löschen von Features muss noch getestet werden. Dazu kann das Example erweitert werden und auf das passende `modifydelete` Event reagieren.

mzur commented:
@rvankevelaer Noch als kurzes Feedback: Dein Code sah soweit gut aus. Ich habe ihn nur noch [ein wenig vereinfacht](https://github.com/openlayers/openlayers/compare/3857f58...d1c0463). Dabei habe ich versucht, den Code von der original Modify Interaction möglichst durchgehend weiterzuverwenden und nur an so wenig Stellen wie möglich zu umgehen oder zu ersetzen. Damit hält man die Wahrscheinlichkeit gering, dass der eigene Code bei einem Update von OpenLayers kaputt geht.

Das einzige was ich noch korrigieren musste war, dass das Sketch Feature von ModifyPolygonBrush auch der `source_` hinzugefügt wurde. Daran sollte aber nichts verändert werden. Stattdessen muss, wie bei PolygonBrush, `overlay_` dafür genutzt werden.

Die Sache mit dem Löschen sehe ich mir noch an. Vermutlich muss das von außerhalb der Interaction geregelt werden.

mzur commented:
Ich nehme noch eine weitere Änderung vor. Und zwar soll das Löschen von Features optional werden. Wenn es deaktiviert ist, kann mit der "subtract" Interaction kein Feature kleiner als der "Brush"-Kreis werden. Das würde ich dann in BIIGLE nutzen, damit man keine Annotationen löschen kann, ohne das nochmal zu bestätigen.

ghost commented:
> Ich nehme noch eine weitere Änderung vor. Und zwar soll das Löschen von Features optional werden. Wenn es deaktiviert ist, kann mit der "subtract" Interaction kein Feature kleiner als der "Brush"-Kreis werden. Das würde ich dann in BIIGLE nutzen, damit man keine Annotationen löschen kann, ohne das nochmal zu bestätigen.

Das klingt gut. Damit wäre das Problem, dass nicht direkt ein Feature gelöscht werden kann, ja auch behoben?

mzur commented:
> Damit wäre das Problem, dass nicht direkt ein Feature gelöscht werden kann, ja auch behoben?

Ja, wobei das bestimmt jetzt auch schon geht, wenn das Feature bei dem Event von Außerhalb entfernt wird.

mzur committed:
Implement removal of features by subtract ModifyPolygonBrush

mzur committed:
Implement allowRemove option for ModifyPolygonBrush

mzur committed:
Simplify polygon difference operation

mzur committed:
Implement direct removal of feature by ModifyPolygonBrush

The draw interaction directly adds features to the source,
so the modify interaction can remove them directly, too.

mzur committed:
Add resizeCondition option for PolygonBrush and ModifyPolygonBrush

mzur commented:
:+1:

ghost commented:
Sehr schön!
